File: backtracking/CombinationSumII.java
//https://leetcode.com/problems/combination-sum-ii/description/
public class CombinationSumII {
    private List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        backtrack(candidates, target, 0, new ArrayList<>(), 0);
        return res;
    }

    private void backtrack(int[] candidates, int target, int i, List<Integer> cur, int total) {
        if (total == target) {
            res.add(new ArrayList<>(cur));
            return;
        }
        if (total > target || i == candidates.length) {
            return;
        }

        cur.add(candidates[i]);
        backtrack(candidates, target, i + 1, cur, total + candidates[i]);
        cur.remove(cur.size() - 1);
        while (i + 1 < candidates.length && candidates[i] == candidates[i + 1]) {
            i++;
        }
        backtrack(candidates, target, i + 1, cur, total);
    }
}

File: backtracking/combination_sum_II.py
# https://leetcode.com/problems/combination-sum-ii/


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def backtrack(i, comb, total):
            if total == target:
                res.append(comb.copy())
                return
            if total > target or i >= len(candidates):
                return

            comb.append(candidates[i])
            backtrack(i + 1, comb, total + candidates[i])
            comb.pop()

            while i + 1 < len(candidates) and candidates[i] == candidates[i + 1]:
                i += 1
            backtrack(i + 1, comb, total)

        backtrack(0, [], 0)
        return res
